File: Example/Tests/RestClient/CraryRestClientMappingTest.m
#import <Specta/Specta.h>
#import <Expecta/Expecta.h>
#import "CraryRestClient.h"
#import "CraryRestClient+Gzip.h"
#import "CraryRestClientAttachment.h"
#import <DCKeyValueObjectMapping/DCKeyValueObjectMapping.h>
#import <DCKeyValueObjectMapping/DCParserConfiguration.h>

#define BASE_URL @"http://localhost:3000/"

@interface CraryRestClientMappingTestPingResult : NSObject
@property (strong, nonatomic) NSString *response;
@end
@implementation CraryRestClientMappingTestPingResult
+ (DCKeyValueObjectMapping *)parser
{
    static DCKeyValueObjectMapping *_parser = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        _parser = [DCKeyValueObjectMapping mapperForClass:[CraryRestClientMappingTestPingResult class]];
    });
    return _parser;
}
@end

@interface CraryRestClientMappingTestDataResult : NSObject
@property (strong, nonatomic) NSString *data;
@end
@implementation CraryRestClientMappingTestDataResult
+ (DCKeyValueObjectMapping *)parser
{
    static DCKeyValueObjectMapping *_parser = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        _parser = [DCKeyValueObjectMapping mapperForClass:[CraryRestClientMappingTestDataResult class]];
    });
    return _parser;
}
@end

@interface CraryRestClientMappingTestTestObject : NSObject
@property (strong, nonatomic) NSString *a;
@property (assign, nonatomic) int b;
@property (assign, nonatomic) BOOL c;
@property (strong, nonatomic) CraryRestClientMappingTestTestObject *d;
@end
@implementation CraryRestClientMappingTestTestObject
+ (DCKeyValueObjectMapping *)parser
{
    static DCKeyValueObjectMapping *_parser = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        _parser = [DCKeyValueObjectMapping mapperForClass:[CraryRestClientMappingTestTestObject class]];
    });
    return _parser;
}
- (id)initA:(NSString *)a b:(int)b c:(BOOL)c d:(CraryRestClientMappingTestTestObject *)d
{
    self = [super init];
    if (self!=nil) {
        self.a = a;
        self.b = b;
        self.c = c;
        self.d = d;
    }
    return self;
}
- (NSDictionary *)toDict
{
    id a = self.a;
    if (a==nil) {
        a = [NSNull null];
    }
    id d = [self.d toDict];
    if (d==nil) {
        d = [NSNull null];
    }
    return @{@"a":a, @"b":@(self.b), @"c":@(self.c), @"d":d};
}
- (void)testA:(NSString *)a b:(int)b c:(BOOL)c d:(CraryRestClientMappingTestTestObject *)d
{
    EXP_expect(self.a).to.equal(a);
    EXP_expect(self.b).to.equal(b);
    EXP_expect(self.c).to.equal(c);
    if (d==nil) {
        EXP_expect(self.d).to.beNil;
    } else {
        EXP_expect(self.d).notTo.beNil;
        [self.d testA:d.a b:d.b c:d.c d:d.d];
    }
}
@end

@interface CraryRestClientMappingTestUnderlineConvertObject : NSObject
@property (assign, nonatomic) int userId;
@property (strong, nonatomic) NSString *fullName;
@property (strong, nonatomic) NSString *phoneNumber;
@property (assign, nonatomic) int id;
@end
@implementation CraryRestClientMappingTestUnderlineConvertObject
+ (DCKeyValueObjectMapping *)parser
{
    static DCKeyValueObjectMapping *_parser = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        _parser = [DCKeyValueObjectMapping mapperForClass:[CraryRestClientMappingTestUnderlineConvertObject class]];
    });
    return _parser;
}
@end

@interface CraryRestClientMappingTestPostAttachmentsResultSub : NSObject
@property (strong, nonatomic) NSString *d;
@property (assign, nonatomic) int e;
@end
@implementation CraryRestClientMappingTestPostAttachmentsResultSub
@end

@interface CraryRestClientMappingTestPostAttachmentsResultFile : NSObject
@property (strong, nonatomic) NSString *fileName;
@property (assign, nonatomic) int size;
@property (strong, nonatomic) NSString *type;
@end
@implementation CraryRestClientMappingTestPostAttachmentsResultFile
@end

@interface CraryRestClientMappingTestPostAttachmentsResult : NSObject
@property (strong, nonatomic) NSString *a;
@property (assign, nonatomic) int b;
@property (strong, nonatomic) CraryRestClientMappingTestPostAttachmentsResultSub *c;
@property (strong, nonatomic) CraryRestClientMappingTestPostAttachmentsResultFile *f1;
@property (strong, nonatomic) CraryRestClientMappingTestPostAttachmentsResultFile *f2;
@end
@implementation CraryRestClientMappingTestPostAttachmentsResult
+ (DCKeyValueObjectMapping *)parser
{
    static DCKeyValueObjectMapping *_parser = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        _parser = [DCKeyValueObjectMapping mapperForClass:[CraryRestClientMappingTestPostAttachmentsResult class]];
    });
    return _parser;
}
@end

@interface CraryRestClientMappingTestDateResult : NSObject
@property (strong, nonatomic) NSDate *d;
@end
@implementation CraryRestClientMappingTestDateResult
+ (DCKeyValueObjectMapping *)parser
{
    static DCKeyValueObjectMapping *_parser = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        DCParserConfiguration *config = [DCParserConfiguration configuration];
        config.datePattern = @"yyyy-MM-dd'T'HH:mm:ss.SSSZ";
        _parser = [DCKeyValueObjectMapping mapperForClass:[CraryRestClientMappingTestDateResult class] andConfiguration:config];
    });
    return _parser;
}
@end

SpecBegin(CraryRestClientMapping)

it(@"GET", ^{
    waitUntil(^(DoneCallback done) {
        CraryRestClient *restClient = [CraryRestClient sharedClient];
        restClient.baseUrl = BASE_URL;
        [restClient get:@"ping" parameters:nil parser:[CraryRestClientMappingTestPingResult parser] complete:^(NSError *error, CraryRestClientMappingTestPingResult *result) {
            EXP_expect(error).to.beNil;
            EXP_expect(result).notTo.beNil;
            EXP_expect(result.response).to.beNil;
            done();
        }];
    });
});

it(@"GET with parameters", ^{
    waitUntil(^(DoneCallback done) {
        CraryRestClient *restClient = [CraryRestClient sharedClient];
        restClient.baseUrl = BASE_URL;
        NSDictionary *parameters = @{@"message": @"hello"};
        [restClient get:@"ping" parameters:parameters parser:[CraryRestClientMappingTestPingResult parser] complete:^(NSError *error, CraryRestClientMappingTestPingResult *result) {
            EXP_expect(error).to.beNil;
            EXP_expect(result).notTo.beNil;
            EXP_expect(result.response).to.equal(@"hello");
            done();
        }];
    });
});

it(@"POST", ^{
    waitUntil(^(DoneCallback done) {
        CraryRestClient *restClient = [CraryRestClient sharedClient];
        restClient.baseUrl = BASE_URL;
        [restClient post:@"ping" parameters:nil parser:[CraryRestClientMappingTestPingResult parser] complete:^(NSError *error, CraryRestClientMappingTestPingResult *result) {
            EXP_expect(error).to.beNil;
            EXP_expect(result).notTo.beNil;
            EXP_expect(result.response).to.beNil;
            done();
        }];
    });
});

it(@"POST with parameters", ^{
    waitUntil(^(DoneCallback done) {
        CraryRestClient *restClient = [CraryRestClient sharedClient];
        restClient.baseUrl = BASE_URL;
        NSDictionary *parameters = @{@"message": @"hello"};
        [restClient post:@"ping" parameters:parameters parser:[CraryRestClientMappingTestPingResult parser] complete:^(NSError *error, CraryRestClientMappingTestPingResult *result) {
            EXP_expect(error).to.beNil;
            EXP_expect(result).notTo.beNil;
            EXP_expect(result.response).to.equal(@"hello");
            done();
        }];
    });
});

it(@"Session", ^{
    waitUntil(^(DoneCallback done) {
        CraryRestClient *restClient = [CraryRestClient sharedClient];
        restClient.baseUrl = BASE_URL;
        NSDictionary *parameters = @{@"data": @"croquis"};
        [restClient post:@"setData" parameters:parameters parser:[CraryRestClientMappingTestDataResult parser] complete:^(NSError *error, CraryRestClientMappingTestDataResult *result) {
            restClient.baseUrl = BASE_URL;
            [restClient get:@"getData" parameters:nil parser:[CraryRestClientMappingTestDataResult parser] complete:^(NSError *error, CraryRestClientMappingTestDataResult *result) {
                EXP_expect(error).to.beNil;
                EXP_expect(result).notTo.beNil;
                EXP_expect(result.data).to.equal(@"croquis");
                done();
            }];
        }];
    });
});

it(@"POST with gzipped parameters", ^{
    waitUntil(^(DoneCallback done) {
        CraryRestClient *restClient = [CraryRestClient sharedClient];
        restClient.baseUrl = BASE_URL;
        NSDictionary *parameters = @{@"message": @"hello"};
        [restClient postGzip:@"ping" parameters:parameters parser:[CraryRestClientMappingTestPingResult parser] complete:^(NSError *error, CraryRestClientMappingTestPingResult *result) {
            EXP_expect(error).to.beNil;
            EXP_expect(result).notTo.beNil;
            EXP_expect(result.response).to.equal(@"hello");
            done();
        }];
    });
});

it(@"Object", ^{
    waitUntil(^(DoneCallback done) {
        CraryRestClient *restClient = [CraryRestClient sharedClient];
        restClient.baseUrl = BASE_URL;
        CraryRestClientMappingTestTestObject *parameters = [[CraryRestClientMappingTestTestObject alloc] initA:@"message" b:5 c:TRUE d:[[CraryRestClientMappingTestTestObject alloc] initA:@"sub" b:0 c:FALSE d:nil]];
        [restClient get:@"echo" parameters:[parameters toDict] parser:[CraryRestClientMappingTestTestObject parser] complete:^(NSError *error, CraryRestClientMappingTestTestObject *result) {
            EXP_expect(error).to.beNil;
            EXP_expect(result).notTo.beNil;
            [result testA:@"message" b:5 c:TRUE d:[[CraryRestClientMappingTestTestObject alloc] initA:@"sub" b:0 c:FALSE d:nil]];
            [restClient post:@"echo" parameters:[parameters toDict] parser:[CraryRestClientMappingTestTestObject parser] complete:^(NSError *error, CraryRestClientMappingTestTestObject *result) {
                EXP_expect(error).to.beNil;
                EXP_expect(result).notTo.beNil;
                [result testA:@"message" b:5 c:TRUE d:[[CraryRestClientMappingTestTestObject alloc] initA:@"sub" b:0 c:FALSE d:nil]];
                done();
            }];
        }];
    });
});

it(@"List", ^{
    waitUntil(^(DoneCallback done) {
        CraryRestClient *restClient = [CraryRestClient sharedClient];
        restClient.baseUrl = BASE_URL;
        CraryRestClientMappingTestTestObject *obj1 = [[CraryRestClientMappingTestTestObject alloc] initA:@"obj1" b:11 c:FALSE d:nil];
        CraryRestClientMappingTestTestObject *obj2 = [[CraryRestClientMappingTestTestObject alloc] initA:@"obj2" b:22 c:TRUE d:[[CraryRestClientMappingTestTestObject alloc] initA:@"sub" b:0 c:FALSE d:nil]];
        CraryRestClientMappingTestTestObject *obj3 = [[CraryRestClientMappingTestTestObject alloc] initA:@"obj3" b:33 c:FALSE d:nil];
        NSArray *parameters = @[[obj1 toDict], [obj2 toDict], [obj3 toDict]];
        [restClient post:@"echo" parameters:parameters parser:[CraryRestClientMappingTestTestObject parser] complete:^(NSError *error, NSArray *result) {
            EXP_expect(error).to.beNil;
            EXP_expect(result).notTo.beNil;
            EXP_expect([result count]).to.equal(3);
            [[result objectAtIndex:0] testA:@"obj1" b:11 c:FALSE d:nil];
            [[result objectAtIndex:1] testA:@"obj2" b:22 c:TRUE d:[[CraryRestClientMappingTestTestObject alloc] initA:@"sub" b:0 c:FALSE d:nil]];
            [[result objectAtIndex:2] testA:@"obj3" b:33 c:FALSE d:nil];
            done();
        }];
    });
});

it(@"Underline/id convert", ^{
    waitUntil(^(DoneCallback done) {
        CraryRestClient *restClient = [CraryRestClient sharedClient];
        restClient.baseUrl = BASE_URL;
        NSDictionary *parameters = @{@"user_id":@(123), @"full_name": @"Crary", @"phone_number":@"1-234-5678", @"id":@(456)};
        [restClient get:@"echo" parameters:parameters parser:[CraryRestClientMappingTestUnderlineConvertObject parser] complete:^(NSError *error, CraryRestClientMappingTestUnderlineConvertObject *result) {
            EXP_expect(error).to.beNil;
            EXP_expect(result).notTo.beNil;
            EXP_expect(result.userId).to.equal(123);
            EXP_expect(result.fullName).to.equal(@"Crary");
            EXP_expect(result.phoneNumber).to.equal(@"1-234-5678");
            EXP_expect(result.id).to.equal(456);
            done();
        }];
    });
});

it(@"POST attachments", ^{
    waitUntil(^(DoneCallback done) {
        CraryRestClient *restClient = [CraryRestClient sharedClient];
        restClient.baseUrl = BASE_URL;
        NSDictionary *parameters = @{@"a": @"message", @"b": @(5), @"c": @{@"d": @"hello", @"e": @(9)}};
        unsigned char file1[] = { 1, 2, 3 };
        CraryRestClientAttachment *attachment1 = [CraryRestClientAttachment newData:[NSData dataWithBytes:file1 length:3] name:@"f1" mimeType:@"image/jpeg" fileName:@"photo.jpg"];
        unsigned char file2[] = { 4, 5, 6, 7, 8, 9, 10 };
        CraryRestClientAttachment *attachment2 = [CraryRestClientAttachment newData:[NSData dataWithBytes:file2 length:7] name:@"f2" mimeType:@"audio/mpeg" fileName:@"sound.mp3"];
        NSArray *attachments = @[attachment1, attachment2];
        [restClient post:@"echo" parameters:parameters attachments:attachments parser:[CraryRestClientMappingTestPostAttachmentsResult parser] complete:^(NSError *error, CraryRestClientMappingTestPostAttachmentsResult *result) {
            EXP_expect(error).to.beNil;
            EXP_expect(result).notTo.beNil;

            EXP_expect(result.a).to.equal(@"message");
            EXP_expect(result.b).to.equal(5);

            EXP_expect(result.c).notTo.beNil;
            EXP_expect(result.c.d).to.equal(@"hello");
            EXP_expect(result.c.e).to.equal(9);
            
            EXP_expect(result.f1).notTo.beNil;
            EXP_expect(result.f1.fileName).to.equal(@"photo.jpg");
            EXP_expect(result.f1.size).to.equal(3);
            EXP_expect(result.f1.type).to.equal(@"image/jpeg");
            
            EXP_expect(result.f2).notTo.beNil;
            EXP_expect(result.f2.fileName).to.equal(@"sound.mp3");
            EXP_expect(result.f2.size).to.equal(7);
            EXP_expect(result.f2.type).to.equal(@"audio/mpeg");

            done();
        }];
    });
});

it(@"Date", ^{
    waitUntil(^(DoneCallback done) {
        CraryRestClient *restClient = [CraryRestClient sharedClient];
        restClient.baseUrl = BASE_URL;

        [restClient post:@"echo" parameters:@{@"d":@"2014-11-25T10:30:05.010Z"} parser:[CraryRestClientMappingTestDateResult parser] complete:^(NSError *error, CraryRestClientMappingTestDateResult *result) {
            EXP_expect(error).to.beNil;
            EXP_expect(result).notTo.beNil;
            NSCalendar *calendar = [[NSCalendar alloc] initWithCalendarIdentifier:NSCalendarIdentifierGregorian];
            calendar.timeZone = [NSTimeZone timeZoneWithAbbreviation:@"UTC"];
            NSDateComponents *components = [calendar components:(NSCalendarUnitYear|NSCalendarUnitMonth|NSCalendarUnitDay|NSCalendarUnitHour|NSCalendarUnitMinute|NSCalendarUnitSecond) fromDate:result.d];
            EXP_expect(components.year).to.equal(2014);
            EXP_expect(components.month).to.equal(11);
            EXP_expect(components.day).to.equal(25);
            EXP_expect(components.hour).to.equal(10);
            EXP_expect(components.minute).to.equal(30);
            EXP_expect(components.second).to.equal(5);
            double time = [result.d timeIntervalSince1970]*1000;
            time = time - floor(time/1000)*1000;
            EXP_expect(time).to.equal(10);
            done();
        }];
    });
});

SpecEnd

File: Example/Tests/RestClient/CraryRestClientMappingTest.swift
import Crary
import ObjectMapper
import XCTest

let BASE_URL = "http://localhost:3000/"

func toInt(_ value: AnyObject?) -> Int? {
    if let value = value {
        if let i = value as? Int {
            return i
        } else if let str = value as? String {
            return Int(str)
        }
    }
    return 0
}

func toBool(_ value: AnyObject?) -> Bool? {
    if let value = value {
        if let i = value as? Bool {
            return i
        } else if let str = value as? String {
            return str == "1"
        }
    }
    return false
}

class CraryRestClientMappingTest: XCTestCase {
    class TestObject: Mappable {
        var a = ""
        var b = 0
        var c = false
        var d: TestObject?

        init(_ a: String, _ b: Int, _ c: Bool, _ d: TestObject?) {
            self.a = a
            self.b = b
            self.c = c
            self.d = d
        }

        required init?(map: Map) {
        }

        func mapping(map: Map) {
            a <- map["a"]
            b <- (map["b"], TransformOf<Int, AnyObject>(fromJSON: { toInt($0) }, toJSON: { $0 as AnyObject }))
            c <- (map["c"], TransformOf<Bool, AnyObject>(fromJSON: { toBool($0) }, toJSON: { $0 as AnyObject }))
            d <- map["d"]
        }

        func toDict() -> NSDictionary {
            var d: AnyObject = NSNull()
            if self.d != nil {
                d = self.d!.toDict()
            }
            return ["a": self.a, "b": self.b, "c": self.c, "d": d] as NSDictionary
        }

        func test(_ a: String, _ b: Int, _ c: Bool, _ d: TestObject?) {
            XCTAssertEqual(self.a, a)
            XCTAssertEqual(self.b, b)
            XCTAssertEqual(self.c, c)
            if d == nil {
                XCTAssertNil(self.d)
            } else {
                XCTAssertNotNil(self.d)
                self.d!.test(d!.a, d!.b, d!.c, d!.d)
            }
        }
    }

    class PostAttachmentsResult: Mappable {
        class Sub: Mappable {
            var d: String?
            var e: Int?

            required init?(map: Map) {
            }

            func mapping(map: Map) {
                d <- map["d"]
                e <- (map["e"], TransformOf<Int, AnyObject>(fromJSON: { toInt($0) }, toJSON: { $0 as AnyObject }))
            }
        }

        class File: Mappable {
            var fileName: String?
            var size: Int?
            var type: String?

            required init?(map: Map) {
            }

            func mapping(map: Map) {
                fileName <- map["file_name"]
                size <- map["size"]
                type <- map["type"]
            }
        }

        var a: String?
        var b: Int?
        var c: Sub?
        var f1: File?
        var f2: File?

        required init?(map: Map) {
        }

        func mapping(map: Map) {
            a <- map["a"]
            b <- (map["b"], TransformOf<Int, AnyObject>(fromJSON: { toInt($0) }, toJSON: { $0 as AnyObject }))
            c <- map["c"]
            f1 <- map["f1"]
            f2 <- map["f2"]
        }
    }

    class DateResult : Mappable {
        var d: Date?

        required init?(map: Map) {
        }

        func mapping(map: Map) {
            d <- (map["d"], CraryDateTransform())
        }
    }

    override func setUp() {
        super.setUp()
    }

    override func tearDown() {
        super.tearDown()
    }

    func testObject() {
        let expectation = self.expectation(description: "")

        let restClient = CraryRestClient.shared()
        restClient.baseUrl = BASE_URL
        let parameters = TestObject("message", 5, true, TestObject("sub", 0, false, nil))
        restClient.get("echo", parameters: parameters.toDict()) { (error, result: TestObject?) -> Void in
            XCTAssertNil(error)
            XCTAssertNotNil(result)
            result!.test("message", 5, true, TestObject("sub", 0, false, nil))

            restClient.post("echo", parameters: parameters.toDict(), complete: { (error, result: TestObject?) -> Void in
                XCTAssertNil(error)
                XCTAssertNotNil(result)
                result!.test("message", 5, true, TestObject("sub", 0, false, nil))

                expectation.fulfill()
            })
        }

        waitForExpectations(timeout: 10) { (error) -> Void in
        }
    }

    func testList() {
        let expectation = self.expectation(description: "")

        let restClient = CraryRestClient.shared()
        restClient.baseUrl = BASE_URL
        let obj1 = TestObject("obj1", 11, false, nil)
        let obj2 = TestObject("obj2", 22, true, TestObject("sub", 0, false, nil))
        let obj3 = TestObject("obj3", 33, false, nil)
        let parameters = [obj1.toDict(), obj2.toDict(), obj3.toDict()] as AnyObject
        restClient.post("echo", parameters: parameters) { (error, result: [TestObject]?) -> Void in
            XCTAssertNil(error)
            XCTAssertNotNil(result)
            XCTAssertEqual(result!.count, 3)
            result![0].test("obj1", 11, false, nil)
            result![1].test("obj2", 22, true, TestObject("sub", 0, false, nil))
            result![2].test("obj3", 33, false, nil)

            expectation.fulfill()
        }

        waitForExpectations(timeout: 10) { (error) -> Void in
        }
    }

    func testPostAttachments() {
        let expectation = self.expectation(description: "")

        let restClient = CraryRestClient.shared()
        restClient.baseUrl = BASE_URL
        let parameters = ["a": "message", "b": 5, "c": ["d": "hello", "e": 9]] as AnyObject
        let file1: [UInt8] = [1, 2, 3]
        let attachment1 = CraryRestClientAttachment(data: Data(bytes: file1), name: "f1", mimeType: "image/jpeg", fileName: "photo.jpg")
        let file2: [UInt8] = [4, 5, 6, 7, 8, 9, 10]
        let attachment2 = CraryRestClientAttachment(data: Data(bytes: file2), name: "f2", mimeType: "audio/mpeg", fileName: "sound.mp3")
        let attachments: [CraryRestClientAttachment] = [attachment1, attachment2]
        restClient.post("echo", parameters: parameters, attachments: attachments) { (error, result: PostAttachmentsResult?) -> Void in
            XCTAssertNil(error)
            XCTAssertNotNil(result)

            XCTAssertEqual(result!.a, "message")
            XCTAssertEqual(result!.b, 5)

            XCTAssertNotNil(result!.c)
            XCTAssertEqual(result!.c!.d, "hello")
            XCTAssertEqual(result!.c!.e, 9)

            XCTAssertNotNil(result!.f1)
            XCTAssertEqual(result!.f1!.fileName, "photo.jpg")
            XCTAssertEqual(result!.f1!.size, 3)
            XCTAssertEqual(result!.f1!.type, "image/jpeg")

            XCTAssertNotNil(result!.f2)
            XCTAssertEqual(result!.f2!.fileName, "sound.mp3")
            XCTAssertEqual(result!.f2!.size, 7)
            XCTAssertEqual(result!.f2!.type, "audio/mpeg")

            expectation.fulfill()
        }

        waitForExpectations(timeout: 10) { (error) -> Void in
        }
    }

    func testDate() {
        let expectation = self.expectation(description: "")

        let restClient = CraryRestClient.shared()
        restClient.baseUrl = BASE_URL
        let parameters = ["d": "2014-11-25T10:30:05.010Z"] as AnyObject
        restClient.post("echo", parameters: parameters) { (error, result: DateResult?) -> Void in
            XCTAssertNil(error)
            XCTAssertNotNil(result)
            let calendar = NSCalendar(calendarIdentifier: NSCalendar.Identifier.gregorian)!
            calendar.timeZone = TimeZone(abbreviation: "UTC")!
            let unitFlags: NSCalendar.Unit = [.year, .month, .day, .hour, .minute, .second]
            let components = calendar.components(unitFlags, from: result!.d!)
            XCTAssertEqual(components.year, 2014)
            XCTAssertEqual(components.month, 11)
            XCTAssertEqual(components.day, 25)
            XCTAssertEqual(components.hour, 10)
            XCTAssertEqual(components.minute, 30)
            XCTAssertEqual(components.second, 5)
            var time = result!.d!.timeIntervalSince1970 * 1000
            time = time - floor(time/1000)*1000
            XCTAssertEqual(time, 10)

            expectation.fulfill()
        }

        waitForExpectations(timeout: 10) { (error) -> Void in
        }
    }
}
